### ISSUE TITLE
Paginate 'contain' does not inherit the same connection of the $query

### ISSUE THREAD

**Issue by celsowm** [cakephp/cakephp#9532]
This is a (multiple allowed):
* [x] bug
* [x] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.3
* Platform and Target: PHP 7 and SQL Server

We have defined a second connection to allow us to use a different database. After that, we have created a TableClass (AcervoTable) for a table and especified the connection name using the static method:
```
public static function defaultConnectionName() {
    	return 'sicaj';
    }
```

and a belongsTo association:

```
$this->belongsTo('Unidade',[
        		'foreignKey'=>'cod_unidade'
        ]);
```

and when I've tried to paginate:

```
$query = TableRegistry::get('Acervo');
$result = $this->paginate($query, ['contain'=>'Unidade']);

```
Cake failed because it tried to find the "unidade" table in the default database and not in the same database/connection of the AcervoTable class

The only way to get the result was create a **UnidadeTable** with the same defaultConnectionName() of **AcervoTable**.

### Expected Behavior
contain inherit the same connection of the $query;
### Actual Behavior
it was necessary to define manually the classes with defaultConnectionName() for each 'contain' association.

**Comment by markstory** [comment]
> it was necessary to define manually the classes with defaultConnectionName() for each 'contain' association.

Yes this is how the ORM works.

**Comment by celsowm** [comment]
Okay Mark, But I hope in the future Cake be able to define the **connection for a group of table**, would be cool.

**Comment by markstory** [comment]
@celsowm How would groups be defined? What would happen where groups overlap or intersect?

**Comment by celsowm** [comment]
@markstory Well, I have some ideas:

1- app.php with "tables" (with exception for repetitions):

```
'sicaj' => [
            'className' => 'Cake\Database\Connection',
            'driver' => 'Cake\Database\Driver\Mysql',
            'persistent' => false,
            'host' => 'localhost',
            //'port' => 'non_standard_port_number',
            'username' => 'my_app',
            'password' => 'secret',
            'database' => 'test_myapp',
            'encoding' => 'utf8',
            'timezone' => 'UTC',
            'cacheMetadata' => true,
            'quoteIdentifiers' => false,
            'log' => true,
            //'init' => ['SET GLOBAL innodb_stats_on_metadata = 0'],
            'url' => env('DATABASE_TEST_URL', null),
            'tables' => ['acervo',unidade']
        ],
```

2- or a new contain parameter:

`$query->contain('Unidade', true, 'sicaj');
`

3- or even a new association parameter:

```
$this->belongsTo('Unidade',[
        		'foreignKey'=>'cod_unidade',
        		'defaultConnectionName'=>'sicaj',
        ]);

```

**Comment by markstory** [comment]
Defining the tables in the connection definition is an interesting idea. I'm not sure including the connection in an association definition would work as two different associations could easily define conflicting connection names.

Is there a reason you don't want to use the existing method to define the default connection?

**Comment by celsowm** [comment]
Well...
I am developing a app in CakePHP but I need to connect to another sqlserver schema (db) of a old and legacy app (made in delphi 5).

The database of this delphi app is a total mess, lot of primary keys in varchar, foreign keys without the "constraint", bizarre table's name and those kind of things. 

One of "products" that I need to get of this database is a thing called acervo judicial (Judicial Collection), and acervo is product made with more than 15 joins + notExists + notIn + some wheres. To make this task a little bit more easy I created a [Table-Valued Function](https://technet.microsoft.com/en-us/library/ms191165(v=sql.105).aspx) to allow us do some reports only doing something like this:

`select field1, field2 from acervo() where ....`

My first attempt on Cake 3 was a direct SQL:

```
$connection = ConnectionManager::get('default');
$results    = $connection->execute('SELECT top 10 * FROM sicaj.dbo.acervo()')->fetchAll('assoc');
```

Even using the default connection, I was able to get some information of sicaj, but unfortunately, a simple array would not be the solution because we need a .ctp with **pagination** and those things ( acervo() has +- 300k lines)

Then, I tried this cool guy:

`$acervos = TableRegistry::get('acervo()')->find();
`

The first problem was not only the different database, but, the (t)sql ORM's query to discover the tables: 
**SqlserverSchema->describeColumnSql()** will never find "**acervo()**" because in the sys.object the "t.name" of acervo() is only acervo (without parentheses). 

Well, with this little dilemma I gave up of using the Table-Valued Function and I recreated the basic part of acervo() using the QueryBuild and unfortunately for each "belongsTo()" it was/is necessary to create all Table Classes because bake was not so useful for this tables without a "real" foreign key.
